README.md
# Donkey Car Competition

## **Background**


**what is Donkey?**

"Donkey is a high level self driving library written in Python. It was developed with a focus on enabling fast experimentation and easy contribution."
 ----> [**Source**](https://docs.donkeycar.com/)

**what is the competition about?**

Donkey Car is a global initiative that allows users to gain access to a powerful RC platform used to integrate a level of autonomy. This year's 2021 competition was hosted online with the usage of a crowd developed platform to test software on and compete.

**Simulator Source**
----> [**Source**](https://github.com/tawnkramer/sdsandbox)

**Gym Donkey Car Source**
----> [**Source**](https://github.com/tawnkramer/gym-donkeycar)

## **Project Demos**

#### Using canny edge detection and the ability to detect the density of white pixels in a given area, the car is able to maneuver around the track.

### Track Demos

###### Warehouse track demo

<img src="https://github.com/AndreSlavescu/Donkey-Car-autonomous-vehicle-projects/blob/main/docs/warehouse-lap.gif?raw=true" width="600px">

### Video Links

###### lap around "warehouse track" using canny edge detection [**here**]()


Donkey-Car-project/HISTORY.rst
=======
History
=======

1.0.0 (2019-07-26)
------------------

* First release on PyPI.

1.0.1 - 1.0.11 (2019-08-04)
-----------------------------

* Testing out deploy system
* Update credits/authors
* flake8 

project-code/test_clientspeed.py
import os
import random
import json
import time
from io import BytesIO
import base64
import logging
import math

from PIL import Image
import numpy as np
from gym_donkeycar.core.sim_client import SDClient

import cv2


###########################################

class SimpleClient(SDClient):

    def __init__(self, address, poll_socket_sleep_time=0.01, verbose=False):
        super().__init__(*address, poll_socket_sleep_time=poll_socket_sleep_time)
        self.last_image = None
        self.car_loaded = False
        self.verbose = verbose
        self.wcCount=0
        self.cmdPos=0
        self.startTime=time.time()
        self.whiteCount = []
        self.prevWC = 0
        # self.cmds = [(225,0,1),(55,0.5,0.5),(15,0.5,0.6),(15,0.3,0.4),(225,0,1)]
        self.cmds = [(320,0,1),(20,-0.22,0.4),(35,0,0.4),(135,1,0.5),(70,0.7,0.5),(735,-0.007,1),(53,0.55,0.8),(40,-0.7,-0.4),(660,-0.05,1),(260,0.1,1)]
        # self.cmds = [(320,0,1),(20,-0.22,0.4),(35,0,0.4),(135,1,0.5),(70,0.7,0.5),(734,-0.007,1),(54,0.5,0.8),(55,-0.7,-0.25),(1,-0.5,1),(460,0,1)]

    def on_msg_recv(self, json_packet):
        if json_packet['msg_type'] == "need_car_config":
            self.send_config()

        if json_packet['msg_type'] == "car_loaded":
            self.send_config()
            self.car_loaded = True
        
        if json_packet['msg_type'] == "telemetry":
            imgString = json_packet["image"]
            image = Image.open(BytesIO(base64.b64decode(imgString)))
            # image.save("camera_a.png")
            # self.last_image = np.asarray(image)
            
            nowTime = time.time()

            # opencvImage = cv2.cvtColor(self.last_image, cv2.COLOR_RGB2BGR)
            opencvImage = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
            self.last_image = opencvImage

            cv2.waitKey(1)
            # opencvImage = numpy.array(image)
            cv2.putText(opencvImage, "Time: "+str(nowTime-self.startTime), (50,50), cv2.FONT_HERSHEY_SIMPLEX, 0.75, (255,255,255), 2)

            cv2.imshow("image_a",opencvImage)

            if self.verbose:
                print("img:", self.last_image.shape)

            # don't have to, but to clean up the print, delete the image string.
            del json_packet["image"]

            if "imageb" in json_packet:
                imgString = json_packet["imageb"]
                image = Image.open(BytesIO(base64.b64decode(imgString)))
                # image.save("camera_b.png")
                # np_img = np.asarray(image)
                opencvImage = cv2.cvtColor(np.array(image), cv2.COLOR_RGB2BGR)
                # opencvImage = numpy.array(image)

                cv2.putText(opencvImage, "Time: "+str(nowTime-self.startTime), (50,50), cv2.FONT_HERSHEY_SIMPLEX, 0.75, (255,255,255), 2)

                cv2.imshow("image_b",opencvImage)
                key = cv2.waitKey(1)
                if key == 27:
                    quit()
                # if self.verbose:
                #     print("imgb:", np_img.shape)

                # don't have to, but to clean up the print, delete the image string.
                del json_packet["imageb"]

            if "lidar" in json_packet:
                lidar = json_packet["lidar"]
                if lidar is not None:
                    if self.verbose:
                        print("lidar:", len(lidar), "pts")

                # don't have to, but to clean up the print, delete the lidar string.
                del json_packet["lidar"]

        if self.verbose:
            print("got:", json_packet)

    def send_config(self):
        '''
        send three config messages to setup car, racer, and camera
        '''
        racer_name = "Your Name"
        car_name = "Car"
        bio = "I race robots."
        country = "Neverland"
        guid = "some random constant string"

        # Racer info
        msg = {'msg_type': 'racer_info',
            'racer_name': racer_name,
            'car_name' : car_name,
            'bio' : bio,
            'country' : country,
            'guid' : guid }
        self.send_now(json.dumps(msg))
        time.sleep(0.2)
        
        # Car config
        # body_style = "donkey" | "bare" | "car01" choice of string
        # body_rgb  = (128, 128, 128) tuple of ints
        # car_name = "string less than 64 char"

        msg = '{ "msg_type" : "car_config", "body_style" : "car01", "body_r" : "255", "body_g" : "0", "body_b" : "255", "car_name" : "%s", "font_size" : "100" }' % (car_name)
        self.send_now(msg)

        # this sleep gives the car time to spawn. Once it's spawned, it's ready for the camera config.
        time.sleep(0.2)

        # Camera config
        # set any field to Zero to get the default camera setting.
        # this will position the camera right above the car, with max fisheye and wide fov
        # this also changes the img output to 255x255x1 ( actually 255x255x3 just all three channels have same value)
        # the offset_x moves camera left/right
        # the offset_y moves camera up/down
        # the offset_z moves camera forward/back
        # with fish_eye_x/y == 0.0 then you get no distortion
        # img_enc can be one of JPG|PNG|TGA        

        # msg = '{ "msg_type" : "cam_config", "fov" : "150", "fish_eye_x" : "1.0", "fish_eye_y" : "0", "img_w" : "640", "img_h" : "480", "img_d" : "1", "img_enc" : "JPG", "offset_x" : "0.0", "offset_y" : "1.0", "offset_z" : "0.0", "rot_x" : "90.0" }'
        # msg = '{ "msg_type" : "cam_config", "fov" : "0", "fish_eye_x" : "0", "fish_eye_y" : "0", "img_w" : "1280", "img_h" : "720", "img_d" : "1", "img_enc" : "JPG", "offset_x" : "0.0", "offset_y" : "0", "offset_z" : "0.0", "rot_x" : "0" }'
        # msg = '{ "msg_type" : "cam_config", "fov" : "0", "fish_eye_x" : "0", "fish_eye_y" : "0", "img_w" : "1280", "img_h" : "720", "img_d" : "3", "img_enc" : "JPG", "offset_x" : "0.0", "offset_y" : "0", "offset_z" : "0.0", "rot_x" : "0" }'
        # msg = '{ "msg_type" : "cam_config", "fov" : "80", "fish_eye_x" : "0", "fish_eye_y" : "0", "img_w" : "640", "img_h" : "480", "img_d" : "1", "img_enc" : "JPG", "offset_x" : "0.0", "offset_y" : "0", "offset_z" : "0.0", "rot_x" : "0" }'
        msg = '{ "msg_type" : "cam_config", "fov" : "0", "fish_eye_x" : "0", "fish_eye_y" : "0", "img_w" : "640", "img_h" : "480", "img_d" : "3", "img_enc" : "JPG", "offset_x" : "0.0", "offset_y" : "0", "offset_z" : "0.0", "rot_x" : "0" }'
        # msg = '{ "msg_type" : "cam_config", "fov" : "130", "fish_eye_x" : "0.0", "fish_eye_y" : "0", "img_w" : "640", "img_h" : "480", "img_d" : "3", "img_enc" : "JPG", "offset_x" : "0.0", "offset_y" : "3.0", "offset_z" : "2.0", "rot_x" : "90.0" }'

        self.send_now(msg)
        time.sleep(0.2)

        # Camera config B, for the second camera
        # set any field to Zero to get the default camera setting.
        # this will position the camera right above the car, with max fisheye and wide fov
        # this also changes the img output to 255x255x1 ( actually 255x255x3 just all three channels have same value)
        # the offset_x moves camera left/right
        # the offset_y moves camera up/down
        # the offset_z moves camera forward/back
        # with fish_eye_x/y == 0.0 then you get no distortion
        # img_enc can be one of JPG|PNG|TGA

        # msg = '{ "msg_type" : "cam_config_b", "fov" : "150", "fish_eye_x" : "1.0", "fish_eye_y" : "1.0", "img_w" : "255", "img_h" : "255", "img_d" : "1", "img_enc" : "JPG", "offset_x" : "3.0", "offset_y" : "3.0", "offset_z" : "0.0", "rot_x" : "90.0" }'
        # self.send_now(msg)
        time.sleep(0.2)

        # Lidar config
        # the offset_x moves camera left/right
        # the offset_y moves camera up/down
        # the offset_z moves camera forward/back
        # degPerSweepInc : as the ray sweeps around, how many degrees does it advance per sample (int)
        # degAngDown : what is the starting angle for the initial sweep compared to the forward vector
        # degAngDelta : what angle change between sweeps
        # numSweepsLevels : how many complete 360 sweeps (int)
        # maxRange : what it max distance we will register a hit
        # noise : what is the scalar on the perlin noise applied to point position
        # Here's some sample settings that similate a more sophisticated lidar:
        # msg = '{ "msg_type" : "lidar_config", "degPerSweepInc" : "2", "degAngDown" : "25", "degAngDelta" : "-1.0", "numSweepsLevels" : "25", "maxRange" : "50.0", "noise" : "0.2", "offset_x" : "0.0", "offset_y" : "1.0", "offset_z" : "1.0", "rot_x" : "0.0" }'
        # And here's some sample settings that similate a simple RpLidar A2 one level horizontal scan.

        # msg = '{ "msg_type" : "lidar_config", "degPerSweepInc" : "2", "degAngDown" : "0", "degAngDelta" : "-1.0", "numSweepsLevels" : "1", "maxRange" : "50.0", "noise" : "0.4", "offset_x" : "0.0", "offset_y" : "0.5", "offset_z" : "0.5", "rot_x" : "0.0" }'
        # self.send_now(msg)
        time.sleep(0.2)


    def send_controls(self, steering, throttle):
        msg = { "msg_type" : "control",
                "steering" : steering.__str__(),
                "throttle" : throttle.__str__(),
                "brake" : "0.0" }
        self.send(json.dumps(msg))

        # this sleep lets the SDClient thread poll our message and send it out.
        time.sleep(self.poll_socket_sleep_sec)

    def update(self):
    #     if self.cmdPos < len(self.cmds):
    #         (cmdCnt,st,th) = self.cmds[self.cmdPos]
    #         if cmdCnt>0:
    #             cmdCnt=cmdCnt-1.5
    #             self.cmds[self.cmdPos] = (cmdCnt,st,th)
    #         else:
    #             self.cmdPos = self.cmdPos+1
    #             if self.cmdPos < len(self.cmds):
    #                 (cmdCnt,st,th) = self.cmds[self.cmdPos]
    #                 cmdCnt=cmdCnt-1.5
    #                 self.cmds[self.cmdPos] = (cmdCnt,st,th)
    #             else:
    #                 #stop the car
    #                 st = 0
    #                 th = 0
        # else:
        #     #stop the car
        #     st = 0
        #     th = 0
        #     quit()
        th = 1
        st = 0
        img = self.last_image
        wc = 0
        try:   
            # crop_img = img[350:350+25, 0:0+639]
            # cv2.imshow("crop1", crop_img)
            crop_img = img[200:200+20, 0:0+639]
            # crop_img = img
            # cv2.imshow("crop2", crop_img)
            img = cv2.Canny(crop_img, 100, 150)
            # ret,img=cv2.threshold(img,133,255,cv2.THRESH_BINARY_INV)
        
            for i in range(5,15):
                x = 305-i*20
                img1 = img[0:20, x:x+20]
                wc = cv2.countNonZero(img1) 
                # cv2.imshow("img1", img1)
                # key = cv2.waitKey(1)
                
                if wc > 20:
                    img[0:20, x:x+20] = 255
                    # if i > 10:
                    #     st = 0.3
                    #     # th = th*0.5
                    # elif i < 5:
                    #     st = -0.3
                        # th = th*0.5
                    print("whiteCount", wc, x)
                    # self.whiteCount[self.wcCount] = wc
                    # self.wcCount = self.wcCount+1
                    wc = i
                    break

            if wc > 0:
                if wc < 5 or wc > 10:
                    th = 0.2
                    if wc < 5:
                        st = -0.7
                    elif wc > 10:
                        st = 0.7
                else:
                    th = 0.4
                    if wc > 5:
                        st = -0.3
                    elif wc < 10:
                        st = 0.3

            # if wc > 0:
            #     if wc < 10:
            #         st = 0.3
            #     elif wc > 5:
            #         st = -0.3
                # if self.prevWC > 0:
                #     if wc > (self.prevWC):
                #         st = 0.2
                #         th = th*0.1
                #         # th = th/abs(wc-self.prevWC)
                #     elif wc < (self.prevWC):
                #         st = -0.2   
                #         th = th*0.1
                #         # th = th/abs(wc-self.prevWC)             
                self.prevWC = wc
            #     print(self.wcCount, self.whiteCount[self.wcCount-1])
                # if self.wcCount > 1 and wc > (self.whiteCount[self.wcCount-2]+5):
                #     st = 0.5
                # elif self.wcCount > 1 and wc < (self.whiteCount[self.wcCount-2]+5):
                #     st = -0.5

            # for i in range(20):
            #     x = 320+i*15
            #     img1 = img[ 0:15, x:x+15]
            #     whiteCount = cv2.countNonZero(img1) 
            #     # cv2.imshow("img1", img1)
            #     # key = cv2.waitKey(1)
            #     print("whiteCount", whiteCount, x)
            #     if whiteCount > 20:
            #         img[0:15, x:x+15] = 255
            #         if i > 15:
            #             st = 0.3
            #             th = 0.3
            #         elif i < 5:
            #             st = -0.3
            #             th = 0.3            

            cv2.imshow("processed", img)
            key = cv2.waitKey(1)
        except Exception as e:
            print(e)
            pass    
        # just random steering now
        # st = 0 #random.random() * 5.0 - 1.0
        # th = 1
        self.send_controls(st, th)

    def update1(self):
        img = self.last_image
        st = 0
        th = 1
        try:   
            # crop_img = img[350:350+25, 0:0+639]
            # cv2.imshow("crop1", crop_img)
            crop_img = img[200:200+25, 0:0+639]
            # cv2.imshow("crop2", crop_img)
            img = cv2.Canny(crop_img, 100, 180)
            cv2.imshow("processed", img)
            key = cv2.waitKey(1)
        except:
            pass    
        self.send_controls(st, th)


###########################################
## Make some clients and have them connect with the simulator

def test_clients():
    logging.basicConfig(level=logging.INFO)

    # test params
    host = "localhost" # "trainmydonkey.com" for virtual racing server
    port = 9091
    num_clients = 1
    clients = []
    time_to_drive = 200


    # Start Clients
    for _ in range(0, num_clients):
        c = SimpleClient(address=(host, port))
        clients.append(c)

    time.sleep(1)

    # Load Scene message. Only one client needs to send the load scene.
    # msg = '{ "msg_type" : "load_scene", "scene_name" : "sparkfun_avc" }'
    msg = '{ "msg_type" : "load_scene", "scene_name" : "generated" }'
    clients[0].send_now(msg)


    # Send random driving controls
    start = time.time()
    do_drive = True
    clients[0].startTime = start
    while time.time() - start < time_to_drive and do_drive:
        for c in clients:
            c.update()
            if c.aborted:
                print("Client socket problem, stopping driving.")
                do_drive = False

    time.sleep(3.0)

    # Exit Scene - optionally..
    # msg = '{ "msg_type" : "exit_scene" }'
    # clients[0].send_now(msg)

    # Close down clients
    print("waiting for msg loop to stop")
    for c in clients:
        c.stop()

    print("clients to stopped")


if __name__ == "__main__":
    test_clients()


Donkey-Car-project/examples/test_cam_config.py
""" Test the gym's code for configuring the DonkeyCar's camera settings.
"""

import os
import argparse
import gym
import gym_donkeycar
import numpy as np
import uuid

if __name__ == "__main__":

    # Initialize the donkey environment
    # where env_name one of:
    env_list = [
        "donkey-warehouse-v0",
        "donkey-generated-roads-v0",
        "donkey-avc-sparkfun-v0",
        "donkey-generated-track-v0",
        "donkey-roboracingleague-track-v0",
        "donkey-waveshare-v0"
    ]

    parser = argparse.ArgumentParser(description='gym_test')
    parser.add_argument('--sim', type=str, default="sim_path",
                        help='path to unity simulator. maybe be left at default if you would like to start the sim on your own.')
    parser.add_argument('--port', type=int, default=9091,
                        help='port to use for websockets')
    parser.add_argument('--env_name', type=str, default='donkey-warehouse-v0',
                        help='name of donkey sim environment', choices=env_list)

    args = parser.parse_args()

#%% SET UP ENVIRONMENT

    cam = (256,256,3)
    
    conf = {"exe_path" : args.sim, 
        "host" : "127.0.0.1",
        "port" : args.port,

        "body_style" : "donkey",
        "body_rgb" : (128, 128, 128),
        "car_name" : "me",
        "font_size" : 100,

        "racer_name" : "test",
        "country" : "USA",
        "bio" : "I am test client",
        "guid" : str(uuid.uuid4()),

        "cam_resolution" : cam,
        "img_w" : cam[0],
        "img_h" : cam[1],
        "img_d" : cam[2],
        }

    env = gym.make(args.env_name, conf=conf)

    print( "Env cam size: {}".format( env.viewer.get_sensor_size() ) )

    speed = 0.5
    steer = 0.0
    max_steer = 1.0

#%% PLAY
    obv = env.reset()
    for t in range(100):
        action = np.array([steer,speed]) # drive straight with small speed
        try:
            obv, reward, done, info = env.step(action)
        except Exception as ex:
            print( "Exception: {}".format( ex ) )

        if obv.shape != cam:
            print( "Invalid Image size: {}".format( obv.shape ) )
        elif t == 10:
            print( "Actual camera size: {}".format( obv.shape ) )

        if done or (info['hit'] is True):
            obv = env.reset()
            print( "Exiting d/h: {}/{}".format( done, info['hit'] ) )
            break

    env.close()


Donkey-Car-project/examples/genetic_alg/simple_gen_driver.py
'''
file: simple_gen_driver.py
author: Tawn Kramer
date: 18 May 2020
notes: a most basic implementation of genetic cross breeding and mutation to attempt to improve
        a neural network. Assumes the standard Keras model from Donkeycar project.
        Lower score means less loss = better.
'''
import os
import random
import time
import argparse
import json
import threading

import gym
import gym_donkeycar
import numpy as np
from PIL import Image

from simple_gen import *

# noisy, noisy tensorflow. we love you.
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
import warnings  
with warnings.catch_warnings():  
    warnings.filterwarnings("ignore",category=FutureWarning)
    import tensorflow as tf
    from tensorflow.keras import backend as K

tf.logging.set_verbosity(tf.logging.ERROR)

# In this environment we just want to reward speed, and allow collisions.
# So we just overload the reward and episode over functions accordingly.
def custom_reward(self, done):
    return self.speed

def custom_episode_over(self):
    self.over = False



class KerasDriveAgent(KerasNNAgent):
    '''
    Attempt to drive better in sim.
    '''

    def __init__(self, model, conf, sess):
        super().__init__(model, conf)
        self.graph = tf.get_default_graph()
        self.sess = sess

    def begin(self):
        self.score = 0.0
        self.th = threading.Thread(target=self.run)
        self.th.setDaemon(True)
        self.th.start()

    def wait(self):
        self.th.join()

    def run(self):
        conf = self.conf
        with self.graph.as_default():
            K.set_session(self.sess)
            self.model._make_predict_function()

        env = gym.make(conf['env_name'], exe_path="remote", host=conf['host'], port=conf['port'])
        
        # setup some custom reward and over functions
        env.set_reward_fn(custom_reward)
        env.set_episode_over_fn(custom_episode_over)

        self.simulate(env)
        env.close()

    def save_model(self, filename):
        with self.graph.as_default():
            K.set_session(self.sess)
            self.model.save(filename)

    def mutate(self):
        with self.graph.as_default():
            K.set_session(self.sess)
            super().mutate()

    def breed(self, agent1, agent2, sess):
        model1, model2 = agent1.model, agent2.model
        jsm = model1.to_json()
        new_model = tf.keras.models.model_from_json(jsm)

        with agent1.graph.as_default():
            K.set_session(agent1.sess)
            w = model1.get_weights()
        
        K.set_session(sess)
        new_model.set_weights(w)

        iLayers = self.conf["layers_to_combine"]
        for iLayer in iLayers:
            layer1 = model1.get_layer(index=iLayer)
            layer2 = model2.get_layer(index=iLayer)
            final_layer = new_model.get_layer(index=iLayer)
            self.merge_layers((final_layer, sess), (layer1, agent1.sess) , (layer2, agent2.sess))

        return new_model

    def merge_layers(self, dest, src1, src2):
        dest_layer, d_sess = dest
        src1_layer, s1_sess = src1
        src2_layer, s2_sess = src2

        K.set_session(s1_sess)
        w1 = src1_layer.get_weights()

        K.set_session(s2_sess)
        w2 = src2_layer.get_weights()

        K.set_session(d_sess)
        res = w1.copy()

        if type(w1) is list:
            half = round(len(w1) / 2)
            res[half:-1] = w2[half:-1]
        else:
            l_indices = np.tril_indices_from(w2)
            res[l_indices] = w2[l_indices]
        dest_layer.set_weights(res)


    def select_action(self, img_arr):
        with self.graph.as_default():
            K.set_session(self.sess)
            one_byte_scale = 1.0 / 255.0
            img_arr = img_arr.astype(np.float32) * one_byte_scale
            img_arr = img_arr.reshape((1,) + img_arr.shape)
            #print("img_arr", type(img_arr), img_arr.shape)
            pred = self.model.predict_on_batch(img_arr)
            steering = pred[0][0][0]
            throttle = pred[1][0][0]
            action = [steering, throttle]
            return action


    def simulate(self, env):

        # Reset the environment
        obv = env.reset()

        for _ in range(self.conf['MAX_TIME_STEPS']):

            # Select an action
            action = self.select_action(obv)

            # execute the action
            obv, reward, done, _ = env.step(action)
            
            self.score += reward

            if done:
                break

        print("agent simulate step done. total reward:", self.score)


    def make_new(self, parent1, parent2):
        config = tf.ConfigProto()
        config.gpu_options.allow_growth = True
        
        sess = tf.Session(config=config)
        K.set_session(sess)

        new_model = self.breed(parent1, parent2, sess)
        agent = KerasDriveAgent(new_model, self.conf, sess)
        agent.mutate()
        return agent



class GeneticPositiveRewardAlg(GeneticAlg):

    def sort_agents(self):
        self.population.sort(key=lambda x: x.get_score(), reverse=True)

    def on_agents_finished(self):
        best_agent = self.population[0]
        best_agent.save_model(self.conf["out_model"])


if __name__ == "__main__":

    # Initialize the donkey environment
    # where env_name one of:
    env_list = [
        "donkey-warehouse-v0",
        "donkey-generated-roads-v0",
        "donkey-avc-sparkfun-v0",
        "donkey-generated-track-v0",
        "donkey-roboracingleague-track-v0",
        "donkey-waveshare-v0"
    ]
	
    parser = argparse.ArgumentParser(description='simple_gen_driver')
    parser.add_argument('--host', type=str, default='127.0.0.1', help='host to use for tcp')
    parser.add_argument('--port', type=int, default=9091, help='port to use for tcp')
    parser.add_argument('--test', action="store_true", help='load the trained model and play')
    parser.add_argument('--num_agents', type=int, default=8, help='how many agents in our population')
    parser.add_argument('--num_iter', type=int, default=8, help='how many generations before we stop')
    parser.add_argument('--max_steps', type=int, default=200, help='how many sim steps before we stop an epoch')
    parser.add_argument('--env_name', type=str, default='donkey-warehouse-v0', help='name of donkey sim environment', choices=env_list)
    parser.add_argument('--in_model', type=str, help='.h5 model produced by donkeycar. expects the default linear model type.')
    parser.add_argument('--out_model', type=str,  default='out.h5', help='.h5 model to save.')

    args = parser.parse_args()

    env_id = args.env_name

    # only needed if TF==1.13.1
    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    

    if args.test:
        pass #todo
        
    else:
    
         # These are the final two dense layers we will mutate. We will use the same two layers we breeding.
        to_mutate = [14, 16]

        conf = { "layers_to_mutate" : to_mutate}
        conf["layers_to_combine"] = to_mutate
        conf["mutation_rate"] = 1.0
        conf["mutation_max"] = 0.1
        conf["mutation_min"] = 0.0
        conf["mutation_decay"] = 1.0
        conf['host'] = args.host
        conf['port'] = args.port
        conf['env_name'] = args.env_name
        conf['MAX_TIME_STEPS'] = args.max_steps
        conf['out_model'] = args.out_model

        population = []
        

        for i in range(args.num_agents):
            sess = tf.Session(config=config)
            K.set_session(sess)
            model = tf.keras.models.load_model(os.path.expanduser(args.in_model))
            agent = KerasDriveAgent(model, conf, sess)
            if i > 0:
                agent.mutate()
            population.append(agent)

        ## Try to improve
        alg = GeneticPositiveRewardAlg(population=population, conf=conf)
        alg.process(num_iter=args.num_iter)

        ## Our best agent
        agent = alg.population[0]
        print("final score:", agent.score)
